### PR TITLE
Initializes plan interfaces for expressions and relations

### PR DESCRIPTION
**Pull request by RCHowell**
## Description

This PR begins defining the plan interfaces which we will use for defining a stable and evolve-able API for the logical algebra. This is a deliberate departure from generated IRs which we are not able to evolve without API breaking changes. Notably we avoid sealed classes/interfaces and only use java-style get/set methods.

I am open to using the kotlin interface values (aka properties) as these become Java-style get methods anyways. It would just be kotlin syntax sugar.

## Other Information
- Updated Unreleased Section in CHANGELOG: **NO**
- Any backward-incompatible changes? **NO**
- Any new external dependencies? **NO**
- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **YES**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/Rex.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.partiql.plan.v1.Node
+import org.partiql.types.PType
+
+/**
+ * TODO DOCUMENTATION
+ */
+public interface Rex : Node {
+
+    public fun getType(): PType
```
Comment (alancai98) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/Rex.kt:11]: nice, I like that the type is on the same level as the `Rex` implementations. Was pretty annoying with the current plan APIs when the `Rex.Op`s couldn't directly reference their `type` field.

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/Statement.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package org.partiql.plan.v1
+
+import org.partiql.plan.v1.rex.Rex
+
+/**
+ * TODO DOCUMENTATION
+ */
+public interface Statement {
```
Comment (alancai98) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/Statement.kt:8]: should `Statement` also implement `Node`?

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/RexCast.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package org.partiql.plan.v1.rex
+
+import org.partiql.types.PType
+
+/**
```
Comment (alancai98) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/RexCast.kt:5]: What's the sequencing for the PRs? Saw that this was marked as 1/2 in the title. Will `Rel`s be in a different PR? Similarly for the visitors + builders?

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/Rex.kt (hunk 1 above)
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/Rex.kt:11]: This is one of the primary goals, on top of being able to have internal implementations of these operators for the typers without having to transform across domains. I am making an effort to flatten things for ease-of-use.

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/RexCast.kt (hunk 3 above)
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/RexCast.kt:5]: Correct, rels will be another **commit**.  Visitors/builders will be another **pull request**.

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelExcludeStep.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package org.partiql.plan.v1.rel
+
+/**
+ * TODO DOCUMENTATION
+ */
+public interface RelExcludeStep : Rel {
```
Comment (alancai98) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelExcludeStep.kt:6]: don't think the `ExcludeStep` needs to implement `Rel`

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelExcludeStep.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    public fun getSteps(): List<RelExcludeStep>
```
Comment (alancai98) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelExcludeStep.kt:8]: nit: could be clearer to use `getSubsteps`

```suggestion
    public fun getSubsteps(): List<RelExcludeStep>
```

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/Rel.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package org.partiql.plan.v1.rel
+
+import org.partiql.plan.v1.Node
+import org.partiql.plan.v1.Schema
+
+public interface Rel : Node {
+
+    public fun getInputs(): List<Rel>
```
Comment (alancai98) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/Rel.kt:8]: So most of the `Rel`s have a `getInput(): Rel` function. Should they actually just use this `getInputs`s function?

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelUnpivot.kt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package org.partiql.plan.v1.rel
+
+import org.partiql.plan.v1.rex.Rex
+
+/**
+ * TODO DOCUMENTATION
+ */
+public interface RelUnpivot : Rel {
+
+    public fun getInput(): Rex
```
[the remark this answers is not in the document]
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelUnpivot.kt:10]: Unpivot does have an input, and the return type is rex.

```
FROM UNPIVOT T
```

Is

```
RelUnpivot(input = RexVar(T))
```

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/RexCall.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ * Scalar function calls.
+ */
+public interface RexCall : Rex {
+
+    /**
+     * Returns the function to invoke.
+     *
+     * @return
+     */
+    public fun getFunction(): String
```
[the remark this answers is not in the document]
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/RexCall.kt:13]: That is correct. The given methods are the **required** ones which we do not change.  Yes we will set defaults for **additional** methods we may introduce. Additionally I am working on abstract bases for extension.

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelScan.kt [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package org.partiql.plan.v1.rel
+
+/**
+ * TODO DOCUMENTATION
+ */
+public interface RelScan : Rel {
+
+    public fun getInput(): Rel
```
[the remark this answers is not in the document]
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelScan.kt:8]: Yes this should be a rex, nice catch

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelExcludeStep.kt (hunk 4 above)
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelExcludeStep.kt:6]: Correct, bad paste

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelExcludeStep.kt (hunk 5 above)
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelExcludeStep.kt:8]: Ack

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/Rel.kt (hunk 6 above)
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/Rel.kt:8]: Some of them have getInput() when it's a single input, and some have getLeft() and getRight() if it's a multi-set operator. This is a top-level getInputs() for tree-traversal

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/Statement.kt [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+    /**
+     * PartiQL Query Statement — i.e. SELECT-FROM
+     */
+    public interface Query {
```
Comment (johnedquinn) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/Statement.kt:13]:
```suggestion
    public interface Query : Statement {
```

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelUnion.kt [hunk 11]
```diff
@@ -0,0 +1,67 @@
+package org.partiql.plan.v1.rel
+
+/**
+ * Logical `UNION [ALL|DISTINCT]` operator for set (or multiset) union.
+ */
+public interface RelUnion : Rel {
+
+    public fun isAll(): Boolean
```
Comment (alancai98) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelUnion.kt:8]: nit: wonder if it would be preferred to keep the set quantifier naming consistent. In `RelUnion`, `RelExcept` and `RelIntersect`, we use `isAll` while in `RelAggregateCall` we use `isDistinct`.

**Review by alancai98 — APPROVED** [review]

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelUnion.kt (hunk 11 above)
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelUnion.kt:8]: This has to do with which is the "default".

For union/intersect/except, DISTINCT is default — so "isAll()" is like asking "is this different than the default"?

For aggregate calls, ALL is default — so "isDistinct()" is like asking "is this different than the default"?
